insecure-time/src/lib.rs

```diff
@@ -0,0 +1,853 @@
+#![no_std]
+#![cfg_attr(all(feature = "std", target_env = "sgx"), feature(sgx_platform))]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+#[cfg(feature = "std")]
+extern crate alloc;
+
+use alloc::borrow::ToOwned;
+
+#[cfg(feature = "std")]
+use std::time::SystemTime;
+
+#[cfg(all(feature = "std", target_os = "linux"))]
+use std::fs;
+
+#[cfg(not(target_env = "sgx"))]
+use core::arch::x86_64::__cpuid;
+use core::arch::x86_64::__rdtscp;
+use core::cell::RefCell;
+use core::cmp::{self, PartialEq, PartialOrd};
+use core::fmt::{self, Debug, Display, Formatter};
+use core::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use core::time::Duration;
+use core::ops::Add;
+
+const NANOS_PER_SEC: u32 = 1_000_000_000;
+
+pub trait NativeTime: PartialOrd + Copy + Add<core::time::Duration, Output = Self> + ToOwned {
+    fn minimum() -> Self;
+    fn abs_diff(&self, other: &Self) -> Duration;
+    fn now() -> Self;
+}
+
+#[cfg(feature = "std")]
+impl NativeTime for SystemTime {
+    fn minimum() -> Self {
+        SystemTime::UNIX_EPOCH
+    }
+
+    fn abs_diff(&self, other: &Self) -> Duration {
+        match self.duration_since(*other) {
+            Ok(duration) => duration,
+            Err(e) => e.duration()
```

Comment:
Just learned about `duration_since`. Since it is odd, I think it merits a comment explaining that `Err(e)` contains the duration when `other` is earlier than self.

Reply:
Added a comment